src/python/gem5/prebuilt/viper/gpu_cache_hierarchy.py

```diff
@@ -0,0 +1,351 @@
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+import math
+
+from m5.objects import (
+    AddrRange,
+    DMASequencer,
+    HBM_2000_4H_1x64,
+    MemCtrl,
+    RubyCache,
+    RubySequencer,
+    RubySystem,
+    SrcClockDomain,
+    TreePLRURP,
+    VIPERCoalescer,
+    VoltageDomain,
+)
+
+from ...coherence_protocol import CoherenceProtocol
+from ...components.cachehierarchies.ruby.abstract_ruby_cache_hierarchy import (
+    AbstractRubyCacheHierarchy,
+)
+from ...components.cachehierarchies.ruby.caches.viper.directory import (
+    ViperGPUDirectory,
+)
+from ...components.cachehierarchies.ruby.caches.viper.dma_controller import (
+    ViperGPUDMAController,
+)
+from ...components.cachehierarchies.ruby.caches.viper.sqc import SQCCache
+from ...components.cachehierarchies.ruby.caches.viper.tcc import TCCCache
+from ...components.cachehierarchies.ruby.caches.viper.tcp import TCPCache
+from ...components.devices.gpus.viper_shader import ViperShader
+from ...utils.requires import requires
+from .viper_network import (
+    SimpleDoubleCrossbar,
+    SimplePt2Pt,
+)
+
+
+class ViperGPUCacheHierarchy(AbstractRubyCacheHierarchy):
+    _seqs = 0
+
+    @classmethod
+    def seqCount(cls):
+        # Use SeqCount not class since we need global count
+        cls._seqs += 1
+        return cls._seqs - 1
+
+    def __init__(
+        self,
+        tcp_size: str,
+        tcp_assoc: int,
+        sqc_size: str,
+        sqc_assoc: int,
+        scalar_size: str,
+        scalar_assoc: int,
+        tcc_size: str,
+        tcc_assoc: int,
+        tcc_count: int,
+        cu_per_sqc: int,
+        num_memory_channels: int,
+        cache_line_size: int,
+        shader: ViperShader,
+    ):
+        """
+        :param size: The size of each cache in the heirarchy.
+        :param assoc: The associativity of each cache.
+        """
+        super().__init__()
+
+        self._tcp_size = tcp_size
+        self._tcp_assoc = tcp_assoc
+        self._sqc_size = sqc_size
+        self._sqc_assoc = sqc_assoc
+        self._scalar_size = scalar_size
+        self._scalar_assoc = scalar_assoc
+        self._tcc_size = tcc_size
+        self._tcc_assoc = tcc_assoc
+        self._cache_line_size = cache_line_size
+
+        self.ruby_gpu = RubySystem()
+        self.ruby_gpu.block_size_bytes = cache_line_size
+
+        # We don't want incorporate_cache which assigns to board, so the rest
+        # of this constructor implements the functionality of incorporate_cache
+        # seen in other stdlib cache hierarchies.
```

Comment:
Can you explain this comment more? I would think we *do* want the board to call `incorporate_cache` with the GPU.

I.e., the `ViperBoard` should override `connect_things()` in that class and call `gpu.incorporate_caches(self)`

Reply:
I changed the comment, but basically there is no reason to delay setting up the device. I haven't looked into the purpose of incorporate_cache but I think it need to wait until the cpu and memory are created first.  This is not the case for the GPU device